influxql/call_iterator.go

```diff
@@ -293,37 +295,46 @@ func stringDistinctReduceSlice(a []StringPoint, opt *reduceOptions) []StringPoin
 func newMeanIterator(input Iterator, opt IteratorOptions) Iterator {
 	switch input := input.(type) {
 	case FloatIterator:
-		return &floatReduceSliceIterator{input: newBufFloatIterator(input), opt: opt, fn: floatMeanReduceSlice}
+		return &floatReduceIterator{input: newBufFloatIterator(input), opt: opt, fn: floatMeanReduce}
 	case IntegerIterator:
-		return &integerReduceSliceFloatIterator{input: newBufIntegerIterator(input), opt: opt, fn: integerMeanReduceSlice}
+		return &integerReduceFloatIterator{input: newBufIntegerIterator(input), opt: opt, fn: integerMeanReduce}
 	default:
 		panic(fmt.Sprintf("unsupported mean iterator type: %T", input))
 	}
 }
 
-// floatMeanReduceSlice returns the mean value within a window.
-func floatMeanReduceSlice(a []FloatPoint, opt *reduceOptions) []FloatPoint {
-	var mean float64
-	var count int
-	for _, p := range a {
-		if math.IsNaN(p.Value) {
-			continue
-		}
-		count++
-		mean += (p.Value - mean) / float64(count)
+// floatMeanReduce returns the mean value within a window.
+func floatMeanReduce(prev, curr *FloatPoint, opt *reduceOptions) (int64, float64, []interface{}) {
+	if prev == nil {
+		return opt.startTime, curr.Value, nil
 	}
-	return []FloatPoint{{Time: opt.startTime, Value: mean}}
+
+	value := prev.Value * float64(prev.Aggregated)
+	if curr.Aggregated > 1 {
+		value += curr.Value * float64(curr.Aggregated)
+		value /= float64(prev.Aggregated + curr.Aggregated)
+	} else {
+		value += curr.Value
+		value /= float64(prev.Aggregated + 1)
+	}
+	return prev.Time, value, prev.Aux
 }
 
 // integerMeanReduceSlice returns the mean value within a window.
-func integerMeanReduceSlice(a []IntegerPoint, opt *reduceOptions) []FloatPoint {
-	var mean float64
-	var count int
-	for _, p := range a {
-		count++
-		mean += (float64(p.Value) - mean) / float64(count)
+func integerMeanReduce(prev *FloatPoint, curr *IntegerPoint, opt *reduceOptions) (int64, float64, []interface{}) {
+	if prev == nil {
```

Comment:
That's an option, but it likely won't affect the code too much. I think it might be easier to keep the current pattern of duplicating the function.

It sucks, but until Go gets some form of templates, it's necessary.